File: nodejs/FlightRadar24/entities/entity.js
const {radians} = require("../util");

const DEFAULT_TEXT = "N/A";

/**
 * Representation of a real entity, at some location.
 */
class Entity {
    /**
     * Constructor of Entity class.
     *
     * @param {number} latitude
     * @param {number} longitude
     */
    constructor(latitude = null, longitude = null) {
        this.__setPosition(latitude, longitude);
    }

    /**
     * @param {number} latitude
     * @param {number} longitude
     */
    __setPosition(latitude, longitude) {
        this.latitude = latitude;
        this.longitude = longitude;
    }

    /**
     * @param {*} info
     * @param {*} [replaceBy]
     * @return {*}
     */
    __getInfo(info, replaceBy = DEFAULT_TEXT) {
        if (info === null || info === undefined || info === DEFAULT_TEXT) return replaceBy;
        return info;
    }

    /**
     * Return the distance from another entity (in kilometers).
     *
     * @param {Entity} entity
     * @return {number}
     */
    getDistanceFrom(entity) {
        if (this.latitude == null || this.longitude == null ||
            entity.latitude == null || entity.longitude == null) {
            throw new Error("Cannot calculate distance: one or both entities have no position.");
        }

        const lat1 = radians(this.latitude);
        const lon1 = radians(this.longitude);
        const lat2 = radians(entity.latitude);
        const lon2 = radians(entity.longitude);

        return Math.acos(Math.sin(lat1) * Math.sin(lat2) + Math.cos(lat1) * Math.cos(lat2) * Math.cos(lon2 - lon1)) * 6371;
    }
}

module.exports = Entity;

File: python/FlightRadar24/entities/entity.py
# -*- coding: utf-8 -*-

from abc import ABC
from math import acos, cos, radians, sin
from typing import Optional


class Entity(ABC):
    """
    Representation of a real entity, at some location.
    """

    _default_text = "N/A"

    def __init__(self, latitude: Optional[float], longitude: Optional[float]):
        """
        Constructor of the Entity class.
        """
        self._set_position(latitude, longitude)

    def _set_position(self, latitude: Optional[float], longitude: Optional[float]) -> None:
        self.latitude = latitude
        self.longitude = longitude

    def get_distance_from(self, entity: "Entity") -> float:
        """
        Return the distance from another entity (in kilometers).
        """
        if (self.latitude is None or self.longitude is None
                or entity.latitude is None or entity.longitude is None):
            raise ValueError("Cannot calculate distance: one or both entities have no position.")

        lat1, lon1 = radians(self.latitude), radians(self.longitude)
        lat2, lon2 = radians(entity.latitude), radians(entity.longitude)

        return acos(sin(lat1) * sin(lat2) + cos(lat1) * cos(lat2) * cos(lon2 - lon1)) * 6371
